File: homework1/deepti_patil/q1_MaxMeanSubArray.py
def max_mean_subarray(nums, k):
    """
    Returns the maximum mean of any contiguous subarray of size k.
    Uses a sliding window to compute sums efficiently in O(n) time.
    """

    # Compute the sum of the first window of size k
    window_sum = sum(nums[:k])
    max_sum = window_sum

    # Slide the window across the array
    for i in range(k, len(nums)):
        # Remove the element leaving the window and add the new one
        window_sum += nums[i] - nums[i - k]

        # Track the maximum sum seen so far
        max_sum = max(max_sum, window_sum)

    # Convert the maximum sum into a mean
    return max_sum / k


# Example usage
print(max_mean_subarray([4, 5, -3, 2, 6, 1], 2))  # 4.5
print(max_mean_subarray([4, 5, -3, 2, 6, 1], 3))  # 3
print(max_mean_subarray([1, 1, 1, 1, -1, -1, 2, -1, -1], 3))  # 1
print(max_mean_subarray([1, 1, 1, 1, -1, -1, 2, -1, -1, 6], 5))  # 1

"""
Time complexity:
- O(n) because we pass through the array once.
Space complexity:
- O(1) extra space.
"""

File: homework1/tom_nguyen/q1_MaxMeanSubArray.cpp
#include <iostream>
#include <vector>
using namespace std;

//uses a sliding window to find the maximum mean of a subarray of size k in O(n)
double maxMeanSubArray(const vector<int>& arr, int k) {
    double windowSum = 0;
    for (int i = 0; i < k; ++i) {
        windowSum += arr[i];
    }
    double maxSum = windowSum;
    for (int i = k; i < (int)arr.size(); ++i) {
        windowSum += arr[i] - arr[i - k];
        if (windowSum > maxSum) {
            maxSum = windowSum;
        }
    }
    return maxSum / k;
}

int main() {
    vector<int> t1 = {4, 5, -3, 2, 6, 1};
    cout << maxMeanSubArray(t1, 2) << endl; // 4.5

    vector<int> t2 = {4, 5, -3, 2, 6, 1};
    cout << maxMeanSubArray(t2, 3) << endl; // 3

    vector<int> t3 = {1, 1, 1, 1, -1, -1, 2, -1, -1};
    cout << maxMeanSubArray(t3, 3) << endl; // 1

    vector<int> t4 = {1, 1, 1, 1, -1, -1, 2, -1, -1, 6};
    cout << maxMeanSubArray(t4, 5) << endl; // 1

    return 0;
}

//time spent: 8 minutes
